### ISSUE TITLE
Infinite loop in the compiler when compiling nested loops

### ISSUE THREAD

**Issue by adabbott** [google/jax#2261]
For my use case, I have a function containing nested for-loops, and the bounds of each loop are dependent on function arguments. 
```python
def f(a,b):
  for i in range(a):
    for j in range(b):
    # ...
```

As-is, a function like this is not jit-compilable, since I cannot pass abstract arguments to Python's `range`. However, I can jit-compile (and use vmap/scan/other awesome JAX stuff) if I use `lax.while_loop`'s with the aid of the `jax.experimental.loops` module for simplicity (thanks, @gnecula !) 

For example, the following simple Python loop,
```python
def control(a,b):                    
  quantity = 0                       
  i = a + b                          
  while i > 0:                       
    j = 2 * i + 1                    
    while j > 0:                     
      k = 2 * j + 1                  
      while k > 0:                   
        quantity += i + j + k        
        k -= 1                       
      j -= 1                         
    i -= 1                           
  return quantity                    
```

can be converted to a JAX-friendly version: 

```python
import jax                        
import jax.numpy as np            
import numpy as onp               
from jax.experimental import loops

def test(a,b):                                                
  with loops.Scope() as s:                                    
    s.quantity = 0.0
    s.a = a
    s.b = b                                          
    s.i = s.a + s.b                                               
    s.j = 2 * s.i + 1                                         
    s.k = 2 * s.j + 1                                         
                                                              
    for _ in s.while_range(lambda: s.i > 0):                  
      s.j = 2 * s.i + 1                                       
      for _ in s.while_range(lambda: s.j > 0):                
        s.k = 2 * s.j + 1                                     
        for _ in s.while_range(lambda: s.k > 0):              
          s.quantity += s.i + s.j + s.k                       
          s.k -= 1                                            
        s.j -= 1                                              
      s.i -= 1                                                
    return s.quantity                                         
```

Since this function `test` is jit-compilable, I can use `vmap` to vectorize it across many possible inputs and get great performance relative to naive python code:

```python
A = np.repeat(np.arange(10), 1000)
B = np.repeat(np.arange(10), 1000)
v_test = jax.vmap(test, (0,0))
%timeit v_test(A,B).block_until_ready()
# 1 loop, best of 3: 951 ms per loop
```
```python
A = onp.repeat(onp.arange(10), 1000)
B = onp.repeat(onp.arange(10), 1000)
%timeit result = [control(A[i],B[i]) for i in range(A.shape[0])]
# 1 loop, best of 3: 25 s per loop
```

### The Issue
The above function `test` works great as written, but I have found several strange bugs when it is expressed in slightly different ways. For example, if the sum in the innermost loop `s.quantity += s.i + s.j + s.k` gets swapped such that `s.k` comes before `s.i` or `s.j`, the loop runs indefinitely, and memory usage continuously climbs until I run out of memory:

```python
def test2(a,b):                                                
  with loops.Scope() as s:                                    
    s.quantity = 0.0
    s.a = a
    s.b = b                                          
    s.i = s.a + s.b                                               
    s.j = 2 * s.i + 1                                         
    s.k = 2 * s.j + 1                                         
                                                              
    for _ in s.while_range(lambda: s.i > 0):                  
      s.j = 2 * s.i + 1                                       
      for _ in s.while_range(lambda: s.j > 0):                
        s.k = 2 * s.j + 1                                     
        for _ in s.while_range(lambda: s.k > 0): 
          #s.quantity += s.i + s.j + s.k  # This works!
          s.quantity += s.i + s.k + s.j   # This runs indefinitely   
          s.k -= 1                                            
        s.j -= 1                                              
      s.i -= 1                                                
    return s.quantity

print(test2(3.,2.))                                       
```

Upon further inspection, I found that this bug only occurs when I _define a loop scope variable which is not used directly in the loop_, in this case `s.a` and `s.b`. The following block of code demonstrates this a little more clearly:

```python
def test3(a,b):
  with loops.Scope() as s:
    s.quantity = 0.0

    s.dummy = 0. # define a scope variable that may or may not participate in the loop

    s.i = a + b
    s.j = 2 * s.i + 1
    s.k = 2 * s.j + 1

    for _ in s.while_range(lambda: s.i > 0):
      s.j = 2 * s.i + 1
      for _ in s.while_range(lambda: s.j > 0):
        s.k = 2 * s.j + 1
        for _ in s.while_range(lambda: s.k > 0):
          #s.quantity += s.i + s.j + s.k            # This works!
          #s.quantity += s.i + s.k + s.j            # Runs indefinitely
          s.quantity += s.i + s.k + s.j + s.dummy   # This works!
          s.k -= 1
        s.j -= 1
      s.i -= 1
    return s.quantity

print(test3(3.,2.))
```

So, we tentatively arrive at the following conclusion:
  * The while-loops do not terminate if you: 
    1. define a loop scope variable which is unused in the loop, and 
    2. do operations in the loop in an order that does not match the nested loop order. In the above example, `s.i + s.j + s.k` works, but `s.i + s.k + s.j` results in an infinite loop. 

This is very unexpected behavior. One can get around it by ensuring that only loops.Scope() variables which are to be used in the loop are defined. However, while investigating this, I found another case where operation order in the loop influences whether or not the loop halts. In this case, defining an _inner-loop_ loops.Scope() variable in terms of function arguments and jit-compiling the function causes a similar issue:

```python
def test4(a,b):
  with loops.Scope() as s:
    s.quantity = 0.0
    s.i = a + b
    s.j = 2 * s.i + 1
    s.k = b + 2

    for _ in s.while_range(lambda: s.i > 0):
      s.j = 2 * s.i + 1
      for _ in s.while_range(lambda: s.j > 0):
        s.k = b + 2 # NOTE this changed! 
        for _ in s.while_range(lambda: s.k > 0):
          #s.quantity += s.i + s.j + s.k # This works!
          s.quantity += s.i + s.k + s.j  # Runs indefinitely only when jit-compiled
          s.k -= 1
        s.j -= 1
      s.i -= 1
    return s.quantity

print(test4(3.,2.))
print(jax.jit(test4)(3.,2.))  # never halts for s.i + s.k + s.j
```

In this case,
  * The nested while-loops never terminate if you:
    1. jit-compile the function, 
    2. define a nested loops.Scope() variable in terms of values outside the loops (`s.k = b + 2`), and 
    3. sum quantities in certain orders (`s.i + s.j + s.k` works, but `s.i + s.k + s.j` runs indefinitely).

These issues apparently do not occur when using `jax.lax.while_loop` itself, but occur due to some bug in `jax.experimental.loops`. I can re-express (rather inconveniently!) the previously defined `control` and `test` functions in terms of purely `jax.lax.while_loop` semantics, and everything works fine: 

```python
def test_laxwhile(a,b):
    quantity = 0.
    i = a + b
    j = 2 * i + 1
    k = 2 * j + 1

    condfun_1 = lambda inp: inp[1] > 0. # i > 0.
    condfun_2 = lambda inp: inp[2] > 0. # j > 0.
    condfun_3 = lambda inp: inp[3] > 0. # k > 0.

    def bodyfun_1(inp):
        quantity, i, j, k = inp
        j = 2 * i + 1
        def bodyfun_2(inp):
            quantity, i, j, k = inp
            k = 2 * j + 1
            def bodyfun_3(inp):
                quantity, i, j, k = inp
                #quantity += i + j + k  # This works!
                quantity += i + k + j   # This works!
                k -= 1.
                return (quantity, i, j, k)
            result = jax.lax.while_loop(condfun_3, bodyfun_3, (quantity,i,j,k))
            quantity = result[0]
            j -= 1.
            return (quantity, i, j, k)
        result = jax.lax.while_loop(condfun_2, bodyfun_2, (quantity,i,j,k))
        quantity = result[0]
        i -= 1.
        return (quantity, i, j, k)

    result = jax.lax.while_loop(condfun_1, bodyfun_1, (quantity,i,j,k))
    return result[0]

print(test_laxwhile(3.,2.))

```

### Summary
The `while_range` iterator from `jax.experimental.loops` results in infinite loops when nested, but only sometimes. The loop termination behavior changes when one alters the operation in the loop to another equivalent operation (e.g., permuting the order of commutative operations such as addition). This is apparently not an issue with `jax.lax.while_loop` itself. It is currently unknown whether this affects other iterators in `jax.experimental.loops` (`range` and `cond_range`).

I am using the latest version of jax (0.1.59) and jaxlib (0.1.39),  and all of the above examples were run on CPU.  Any help or insight would be greatly appreciated.  Thanks for reading!

**Comment by gnecula** [comment]
I am confirming the bug for now, I am looking into it. Very strange. The loops are looping more than intended :-)

**Comment by gnecula** [comment]
The best guess so far is that this is a compiler bug. It is not a bug in the loops syntactic sugar, I was able to repro with just `lax.while_loop` (but you must have the extra dummy state, unlike in your attempt). I have not played with `jit`, but I am not surprised it happens also. In JAX, when you run a `lax.while_loop` we compile it anyway, whether or not you asked for `jit`.

I filed the bug with the Google compiler team, I hope they can fix it quickly. 

I can only imagine how much time it took you to narrow this down. Thank you!

**Comment by adabbott** [comment]
Thanks for looking into this! Apologies for the misdiagnosis, I see now I made a mistake in the 'lax.while_loop' repro.  Looking forward to a fix, thanks!

**Comment by hellcats** [comment]
I also am seeing lax.while_loop go into an infinite loop.  This is part of an iterative solver where the convergence criteria is used as the while_loop cond().  I've replaced the while_loop with a fori_loop using a fixed number of iterations, but of course this isn't ideal (it might either do extra work, or fail to fully converge). If I get a chance I'll try reducing the complexity to something I can post.

**Comment by gnecula** [comment]
This has been confirmed as an XLA compiler bug, but one not easy to fix :-( It can affect any loops, whether written with lax.fori_loop, lax.while_loop, or lax.scan (they all lower to XLA loops). I'll keep this bug open, and monitor the situation.

**Comment by gnecula** [comment]
This bug has been fixed in the XLA compiler. You can try it if you build from source, or wait for the next jaxlib release.

**Comment by gnecula** [comment]
I have verified this is fixed in jaxlib 0.1.42
